src/scheduler.rs

```diff
@@ -0,0 +1,391 @@
+    /// [ProposalPsbt].
+    fn pop_payjoin(
+        &self,
+        mut psbt: PartiallySignedTransaction,
+    ) -> Result<Option<ProposalPsbt>, SchedulerError> {
+        let mut pj_by_spk = self.pjs.lock().unwrap();
+
+        // remove signatures from psbt
+        psbt.unsigned_tx.input.iter_mut().for_each(|txin| txin.script_sig = bitcoin::Script::new());
+
+        // find vout of owned output, pop scheduled psbt
+        let proposed_psbt = psbt
+            .unsigned_tx
+            .output
+            .iter()
+            .enumerate()
+            .find_map(|(vout, txout)| pj_by_spk.remove(&txout.script_pubkey).map(|pj| (vout, pj)))
+            .map(|(owned_vout, pj)| ProposalPsbt { psbt, owned_vout, pj });
+
+        // check psbt amount
+        // TODO: we still pop scheduled pj if this happens, is this okay?
+        if let Some(psbt) = proposed_psbt.as_ref() {
+            if !psbt.check_amounts() {
+                return Err(SchedulerError::OriginalPsbtInvalidAmount);
+            }
+        }
+
+        Ok(proposed_psbt)
+    }
+}
+
+pub struct ProposalPsbt {
+    psbt: PartiallySignedTransaction,
+    owned_vout: usize,
+    pj: ScheduledPayJoin,
+}
+
+impl ProposalPsbt {
+    pub fn owned_output(&self) -> &TxOut { &self.psbt.unsigned_tx.output[self.owned_vout] }
+
+    pub fn owned_output_mut(&mut self) -> &mut TxOut {
+        &mut self.psbt.unsigned_tx.output[self.owned_vout]
+    }
+
+    pub fn check_amounts(&self) -> bool {
+        let chan_amounts_sum = self.pj.sum_channel_amounts();
+        let fees = self.pj.calculate_fees();
+        let wallet_amount = self.pj.wallet_amount();
+
+        let owned_txout_value = self.owned_output().value;
+
+        (chan_amounts_sum + fees + wallet_amount).as_sat() == owned_txout_value
+    }
+
+    pub async fn multi_open_channel(
+        &self,
+        lnd: &LndClient,
+    ) -> Result<(Vec<TxOut>, Vec<ChannelId>), SchedulerError> {
+        let requests = self.pj.generate_open_channel_requests();
+
+        let handles = requests
+            .iter()
+            .cloned()
+            .map(|(node, _, req)| {
+                let lnd = lnd.clone();
+                tokio::spawn(async move { open_channel(lnd, node, req).await })
+            })
+            .collect::<Vec<_>>();
+
+        let mut funding_vouts = Vec::<TxOut>::with_capacity(handles.len());
+        for handle in handles {
+            match handle.await.unwrap()? {
+                Some(vout) => funding_vouts.push(vout),
+                None => eprintln!("failed to receive funding psbt after channel open request - this is not handled"),
+            };
+        }
+
+        if funding_vouts.is_empty() {
+            Err(SchedulerError::PayJoinCannotOpenAnyChannel)
+        } else {
+            let chan_ids = requests.iter().map(|(_, id, _)| *id).collect::<Vec<_>>();
+            Ok((funding_vouts, chan_ids))
+        }
+    }
```

Comment:
Multi-channel open is now done async.

Reply:
Receiver negotiates channel openings in parallel in stead of one blocking the next